node/core/candidate-selection/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+		match msg {
+			AllMessages::CandidateValidation(msg) => Ok(FromJob::Validation(msg)),
+			AllMessages::CandidateBacking(msg) => Ok(FromJob::Backing(msg)),
+			AllMessages::CollatorProtocol(msg) => Ok(FromJob::Collator(msg)),
+			_ => Err(()),
+		}
+	}
+}
+
+#[derive(Debug, derive_more::From)]
+enum Error {
+	#[from]
+	Sending(mpsc::SendError),
+	#[from]
+	Util(util::Error),
+	#[from]
+	OneshotRecv(oneshot::Canceled),
+	#[from]
+	ChainApi(ChainApiError),
+	#[from]
+	Runtime(RuntimeApiError),
+}
+
+impl JobTrait for CandidateSelectionJob {
+	type ToJob = ToJob;
+	type FromJob = FromJob;
+	type Error = Error;
+	type RunArgs = ();
+	type Metrics = Metrics;
+
+	const NAME: &'static str = "CandidateSelectionJob";
+
+	/// Run a job for the parent block indicated
+	//
+	// this function is in charge of creating and executing the job's main loop
+	fn run(
+		_relay_parent: Hash,
+		_run_args: Self::RunArgs,
+		metrics: Self::Metrics,
+		receiver: mpsc::Receiver<ToJob>,
+		sender: mpsc::Sender<FromJob>,
+	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>> {
+		async move {
+			let job = CandidateSelectionJob::new(metrics, sender, receiver);
+
+			// it isn't necessary to break run_loop into its own function,
+			// but it's convenient to separate the concerns in this way
+			job.run_loop().await
+		}
+		.boxed()
+	}
+}
+
+impl CandidateSelectionJob {
+	pub fn new(
+		metrics: Metrics,
+		sender: mpsc::Sender<FromJob>,
+		receiver: mpsc::Receiver<ToJob>,
+	) -> Self {
+		Self {
+			sender,
+			receiver,
+			metrics,
+			seconded_candidate: None,
+		}
+	}
+
+	async fn run_loop(mut self) -> Result<(), Error> {
+		while let Some(msg) = self.receiver.next().await {
+			match msg {
+				ToJob::CandidateSelection(CandidateSelectionMessage::Collation(
+					relay_parent,
+					para_id,
+					collator_id,
+				)) => {
+					self.handle_collation(relay_parent, para_id, collator_id)
+						.await;
+				}
+				ToJob::CandidateSelection(CandidateSelectionMessage::Invalid(
+					_,
+					candidate_receipt,
+				)) => {
+					self.handle_invalid(candidate_receipt).await;
+				}
+				ToJob::Stop => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	async fn handle_collation(
+		&mut self,
+		relay_parent: Hash,
+		para_id: ParaId,
+		collator_id: CollatorId,
+	) {
+		if self.seconded_candidate.is_none() {
+			let (candidate_receipt, pov) = match get_collation(
+				relay_parent,
+				para_id,
+				self.sender.clone(),
+			)
+			.await
+			{
+				Ok(response) => response,
+				Err(err) => {
+					log::warn!(target: "candidate_selection", "failed to get collation from collator protocol subsystem: {:?}", err);
+					return;
+				}
+			};
+
+			let pov = Arc::new(pov);
+
+			if !candidate_is_valid(
+				candidate_receipt.descriptor.clone(),
+				pov.clone(),
+				self.sender.clone(),
+			)
+			.await
+			{
+				return;
+			}
+
+			let pov = Arc::try_unwrap(pov)
+				.expect("only clone went to fn candidate_is_valid, which is complete; qed");
```

Comment:
This is kind of shaky. Ownership of the PoV has been transferred to the candidate validation subsystem, and we don't know what it's done with it in this section of the code.